### ISSUE TITLE
"Unknown terminal type" on startup

### ISSUE THREAD

**Issue by ro6** [urbit/urbit#1960]
I downloaded the latest Linux binary and succeeded in creating a comet called `test-comet`, but eventually the startup process failed with:
```
~/urbit-linux64-v0.9.2 ./urbit test-comet
~
urbit 0.9.2
boot: home is /home/me/urbit-linux64-v0.9.2/test-comet
loom: mapped 2048MB
lite: arvo formula 52fd6bb6
lite: core 6ce08503
lite: final state 6ce08503
loom: mapped 2048MB
boot: protected loom
live: logical boot
boot: installed 262 jets
'rxvt-unicode': unknown terminal type.
```

I'm not really sure how `terminfo` stuff works, but I do see the `urbit-terminfo` directory, and the fact that `urbit-terminfo/r/rxvt-unicode` doesn't exist.

**Comment by ro6** [comment]
I was able to work around this issue by copying my local machine's `rxvt-unicode` terminfo file into the matching location in `urbit-terminfo`, so:
```
cp /usr/lib/terminfo/r/rxvt-unicode ~/urbit-linux64-v0.9.2/urbit-terminfo/r/
```
on Arch Linux.

Then I ran into an assertion error documented in #1961.

**Comment by syzygyzer** [comment]
I'm running into what seems to be the same issue or closely related.

After updating to 0.10.0, I am first able to successfully build a fake zod:

```
[poprox@syzygyzer:~/Dropbox/urbit]$ tar xzf urbit-linux64-v0.10.0.tgz                    
[poprox@syzygyzer:~/Dropbox/urbit]$ ./urbit -F zod                   
~
urbit 0.10.0
boot: home is /home/poprox/Dropbox/urbit/zod
loom: mapped 2048MB
lite: arvo formula 7c50e147
lite: core 6c9faaf4
lite: final state 6c9faaf4
boot: downloading pill https://bootstrap.urbit.org/urbit-v0.10.0.pill
loom: mapped 2048MB
boot: protected loom
live: logical boot
boot: installed 262 jets
work: play 1
boot: ship: ~zod (fake)
work: (1)| boot
work: (2)| boot
work: (3)| boot
work: (4)| boot
work: (5)| boot
work: (5)| pill: 2a4f7637
1-b
...
sync succeeded from %base on ~zod to %kids
~zod:dojo> 
```

But the subsequent boot results in the "Unknown terminal type" error.

```
[poprox@syzygyzer:~/Dropbox/urbit]$ ./urbit zod
~
urbit 0.10.0
boot: home is /home/poprox/Dropbox/urbit/zod
loom: mapped 2048MB
lite: arvo formula 7c50e147
lite: core 6c9faaf4
lite: final state 6c9faaf4
loom: mapped 2048MB
boot: protected loom
live: loaded: MB/81.575.936
boot: installed 262 jets
'xterm-256color': unknown terminal type.

[poprox@syzygyzer:~/Dropbox/urbit]$ 
```

Any subsequent attempts to build a ship also fail:

```
[poprox@syzygyzer:~/Dropbox/urbit]$ ./urbit -F litzod
~
urbit 0.10.0
boot: home is /home/poprox/Dropbox/urbit/litzod
loom: mapped 2048MB
lite: arvo formula 7c50e147
lite: core 6c9faaf4
lite: final state 6c9faaf4
boot: downloading pill https://bootstrap.urbit.org/urbit-v0.10.0.pill
loom: mapped 2048MB
boot: protected loom
live: logical boot
boot: installed 262 jets
'xterm-256color': unknown terminal type.
```

**Comment by syzygyzer** [comment]
Alright, after installing v0.10.1 I am no longer encountering this issue.

**Comment by syzygyzer** [comment]
Issue is back with v0.10.3!

I performed a personal breach on my planet (since it had a `bail: meme` error for the past couple weeks) and just tried to launch my planet with the new key, and got this:  

```
./urbit -w tidrel-tabpub -k tidrel-tabpub-9.key
~
urbit 0.10.3
boot: home is /home/poprox/Dropbox/urbit/tidrel-tabpub
loom: mapped 2048MB
lite: arvo formula 69dfe787
lite: core 31c5cfd7
lite: final state 31c5cfd7
boot: downloading pill https://bootstrap.urbit.org/urbit-v0.10.3.pill
boot: retrieving latest block
boot: retrieving ~tidrel-tabpub's public keys
boot: verifying keys
boot: getting sponsor
boot: retrieving galaxy table
boot: retrieving network domains
boot: retrieving keys for sponsor ~litzod
boot: retrieving keys for sponsor ~zod
loom: mapped 2048MB
boot: protected loom
live: logical boot
boot: installed 262 jets
'xterm-256color': unknown terminal type.
```

I get the same issue trying to build a fake zod:

```
./urbit -F zod
~
urbit 0.10.3
boot: home is /home/poprox/Dropbox/urbit/zod
loom: mapped 2048MB
lite: arvo formula 69dfe787
lite: core 31c5cfd7
lite: final state 31c5cfd7
boot: downloading pill https://bootstrap.urbit.org/urbit-v0.10.3.pill
loom: mapped 2048MB
boot: protected loom
live: logical boot
boot: installed 262 jets
'xterm-256color': unknown terminal type.
```
Attempting to mine a comet now, will update when that finishes.

**Comment by syzygyzer** [comment]
Ok, v10.0.2 builds a fake zod fine. So this was definitely reintroduced in the v10.0.2->v10.0.3 upgrade.  

As before, I am running NixOS 19.09, running bash in Konsole.

**Comment by jtobin** [comment]
Huh, I can't reproduce on NixOS 20.03.  I tried booting a fake ship, booting a real ship, and resuming a booted ship -- all work.

I did make some changes to the release packaging in 5918bd1fa, and I did notice a [strange symbolic link error](https://github.com/urbit/urbit/pull/2217#issuecomment-580585680) when testing that procedure on my macOS machine, but I couldn't reproduce it elsewhere, and particularly not in CI (which builds the release binaries).

I don't understand the terminfo database well, but there may be some issue with how symbolic links are handled in it -- that they're fragile when moved about, or between different machines, or something like that.

The release packaging hadn't changed since v0.8.0, so I presume whatever problem you're encountering has really existed on all binaries since then.

**Comment by syzygyzer** [comment]
I'm behind on some other stuff but I'll try running it in 20.03 next week and get back to you. If there's anything other info I can provide, let me know.

**Comment by syzygyzer** [comment]
You are definitely correct about things being very fragile, @jtobin.

I deleted and reinstalled v10.0.2 and built another fake zod. Closing out with C-d and then reopening results in the unknown terminal type error.

I also deleted and reinstalled v10.0.3 and built a fake zod without first trying to launch my planet as I did before. Again, it built fine, but then C-d and reopening results in the unknown terminal type error. 

In short, it works the first time, then results in an error. 

This is still using NixOS 19.09. Not sure if I will get around to testing with unstable this week.

**Comment by syzygyzer** [comment]
I decided to try out another terminal emulator, rxvt_unicode aka urxvt.

I built a fake zod in urxvt, booted it successfully, then closed out. Next I opened Konsole and attempted to boot the same fake zod, and got the `unknown terminal type` error. I went back to urxvt and successfully booted the same fake zod. This is still under NixOS 19.09 and bash.

I don't know if this test is important or not, but I was curious as to why Konsole will let me boot a ship the first time and not a second time, and wondered if it was polluting terminfo somehow when it first runs. If it does pollute it, it is in a way that appears to be isolated from other terminal emulators.

**Comment by justinmk** [comment]
@syzygyzer no need to change the actual terminal emulator, just changing the TERM env var should be enough.

**Comment by syzygyzer** [comment]
Well this bug has arisen again. I have an inkling that this may be more of a NixOS issue than an urbit issue, but I figure I should share here anyways. It seems more severe than last time, since it happening for both Konsole and urxvt on my system, and I tested half a dozen random others by changing the `TERM` environmental variable as @justinmk suggested.

This morning I updated from NixOS 19.09 to 20.03. This included some changes to how the system sets up consoles - they renamed the options used to set console defaults (though I don't know if they changed anything other than the name yet). Changing the name could be sufficient to break things anyways, of course, if this whole apparatus is that fragile.

I updated my `configuration.nix` to reflect the changes to the option name before running the update. Afterward, I get the following when I try to boot my planet, boot an already-built fake zod, or create a new fake zod:

```
$ ./urbit tidrel-tabpub/
~
urbit 0.10.4
boot: home is /home/poprox/Dropbox/urbit-v0.10.4-linux64/tidrel-tabpub
loom: mapped 2048MB
lite: arvo formula 56587b84
lite: core 715fdde3
lite: final state 715fdde3
loom: mapped 2048MB
'xterm-new': unknown terminal type.
boot: protected loom
live: loaded: MB/871.841.792
boot: installed 262 jets
work: fail: broken pipe
```

I tried switching back to the deprecated NixOS console options that I was using before. This did not change anything.

So I tried redownloading the `v0.10.4` binary, in case the `urbit-terminfo` folder gets changed over time (I don't know how this folder works yet). This did change things - I was able to build a fake zod and enter dojo with multiple different values of `TERM`, but subsequent boots end with `work: fail: end of file` instead.

```
$ ./urbit zod
~
urbit 0.10.4
boot: home is /home/poprox/Dropbox/urbit-test/urbit-v0.10.4-linux64/zod
loom: mapped 2048MB
lite: arvo formula 56587b84
lite: core 715fdde3
lite: final state 715fdde3
loom: mapped 2048MB
boot: protected loom
live: loaded: MB/100.024.320
boot: installed 262 jets
'xterm-256color': unknown terminal type.
work: fail: end of file
```

Then I tried restoring the new console options, and tried everything again. The old `v0.10.4` urbit gives the `work: fail: broken pipe` message, while the new `v0.10.4` urbit gives the `work: fail: end of file` message. Making fake zods remains impossible.

I should also mention that every time I reran `nixos-rebuild switch`, I would boot up a new terminal before running any more commands. I have not been restarting my computer between each attempt though (which I don't think is necessary here, but I don't know deep terminal lore).

Pinging @jtobin

**Comment by syzygyzer** [comment]
Someone commented in urbit-help the other day that compiling the binary yourself will make it use your system's own `terminfo`. I did so and finally I can run urbit in NixOS again! The days of creating a new fake zod every single day are now behind me (as it would always run the first time, the error would just arise when I tried to resume it later).

**Comment by joemfb** [comment]
This is fixed in #3064.
